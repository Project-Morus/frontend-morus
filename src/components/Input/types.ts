import { InputHTMLAttributes } from "react";

export interface IInputProps extends InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  helperText?: string;
  hasError?: boolean;
}
