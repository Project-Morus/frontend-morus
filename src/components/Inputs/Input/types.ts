import { InputHTMLAttributes } from "react";
import { IconNamesTypes } from "../../DataDisplay/Icon/types";

export interface IInputProps extends InputHTMLAttributes<HTMLInputElement> {
  id: string;
  label: string;
  errorText?: string;
  hasError?: boolean;
  maxWidth?: number;
  iconName?: IconNamesTypes;
  onClick?: () => void;
}
