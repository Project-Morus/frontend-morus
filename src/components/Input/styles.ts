import styled, { css } from "styled-components";

export const ContainerSC = styled.div<{ $maxWidth?: number }>`
  display: flex;
  flex-direction: column;
  width: 100%;

  & + & {
    margin-top: ${({ theme }) => theme.spacing[5]};
  }

  ${({ $maxWidth }) =>
    $maxWidth &&
    css`
      max-width: ${$maxWidth}px;
    `}
`;

export const LabelSC = styled.label<{ disabled?: boolean }>`
  color: ${({ theme }) => theme.colors.grey[400]};
  margin-bottom: ${({ theme }) => theme.spacing[2]};

  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.colors.grey[100]};
    `}
`;

export const InputSC = styled.input<{ hasError?: boolean }>`
  color: ${({ theme }) => theme.colors.grey[400]};
  padding: ${({ theme }) => theme.spacing[3]};
  border: ${({ theme }) => theme.border.width.regular} solid ${({ theme }) => theme.colors.grey[100]} ;

  border-radius: ${({ theme }) => theme.border.radius.regular};

  &:focus {
    border-color: ${({ theme }) => theme.colors.primary[500]};
    color: ${({ theme }) => theme.colors.primary[500]};
  }

  &::placeholder  {
    color: ${({ theme }) => theme.colors.grey[50]};
  }

  &:disabled {
      color: ${({ theme }) => theme.colors.grey[50]};
      border-color: ${({ theme }) => theme.colors.grey[50]};
      background: ${({ theme }) => theme.colors.white[50]};
      cursor: not-allowed;
      opacity: 0.5;
    }

  ${({ hasError }) =>
    hasError &&
    css`
      color: ${({ theme }) => theme.colors.red[500]};
      border-color: ${({ theme }) => theme.colors.red[500]};
    `}
`;

export const ErrorTextSC = styled.p`
  color: ${({ theme }) => theme.colors.red[500]};
  font-size: ${({ theme }) => theme.typography.TEXT_SIZES.xsmall};
  margin-top: ${({ theme }) => theme.spacing[3]};
`;
