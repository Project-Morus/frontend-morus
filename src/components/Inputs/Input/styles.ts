import styled, { css } from "styled-components";
import Icon from "../../DataDisplay/Icon";

export const ContainerSC = styled.div<{ $maxWidth?: number }>`
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  ${({ $maxWidth }) =>
    $maxWidth &&
    css`
      max-width: ${$maxWidth}px;
    `}
`;

export const LabelSC = styled.label<{ $hasError?: boolean; $disabled?: boolean }>`
  color: ${({ theme, $hasError }) => ($hasError ? theme.colors.red[500] : theme.colors.grey[400])};
  margin-bottom: ${({ theme }) => theme.spacing[2]};

  ${({ $disabled }) =>
    $disabled &&
    css`
      color: ${({ theme }) => theme.colors.grey[100]};
    `}
`;

export const ContentSC = styled.div`
position: relative;
width: 100%;
gap: ${({ theme }) => theme.spacing[4]};
`;

export const InputSC = styled.input<{ $hasError?: boolean; $iconName?: boolean }>`
width: 100%;
color: ${({ theme }) => theme.colors.grey[400]};
padding: ${({ theme }) => theme.spacing[3]};
border: ${({ theme }) => theme.border.width.regular} solid ${({ theme }) => theme.colors.grey[100]};

border-radius: ${({ theme }) => theme.border.radius.regular};

  &:focus {
  border-color: ${({ theme }) => theme.colors.primary[500]};
  color: ${({ theme }) => theme.colors.primary[500]};
}

  &::placeholder {
  color: ${({ theme }) => theme.colors.grey[50]};
}

  &:disabled {
  color: ${({ theme }) => theme.colors.grey[50]};
  border-color: ${({ theme }) => theme.colors.grey[50]};
  background: ${({ theme }) => theme.colors.white[50]};
  cursor: not-allowed;
  opacity: 0.5;
}

  ${({ $hasError }) =>
    $hasError &&
    css`
      color: ${({ theme }) => theme.colors.red[500]};
      border-color: ${({ theme }) => theme.colors.red[500]};
    `}
`;

export const CustomIconSC = styled(Icon)`
  position: absolute;
  top: 8px;
  right: 8px;
`;

export const ErrorTextSC = styled.p`
  color: ${({ theme }) => theme.colors.red[500]};
  font-size: ${({ theme }) => theme.typography.TEXT_SIZES.xsmall};
  margin-top: ${({ theme }) => theme.spacing[3]};
`;
