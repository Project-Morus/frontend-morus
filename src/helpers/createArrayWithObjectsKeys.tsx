export const createArrayWithObjectsKeys = (quantity: number) => {
  const dotsKeys = []

  for (let i = 1; i <= quantity; i++) {
    [...dotsKeys, dotsKeys.push({ key: i })]
  }

  return dotsKeys
}