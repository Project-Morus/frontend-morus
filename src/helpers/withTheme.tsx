import { ComponentType, JSX } from 'react';
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "../styles/themes/default";

type ComponentPropsType<T> = T & JSX.IntrinsicAttributes

export function withTheme<T>(WrapperComponent: ComponentType<ComponentPropsType<T>>) {
  return (props: ComponentPropsType<T>) => {
    return (
      <ThemeProvider theme={defaultTheme}>
        <WrapperComponent {...props} />
      </ThemeProvider>
    );
  };
}

