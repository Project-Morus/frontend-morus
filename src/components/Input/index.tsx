import { forwardRef, useId } from "react";
import { ContainerSC, InputSC, LabelSC, HelperTextSC } from "./styles";
import { IInputProps } from "./types";

const Input = forwardRef<HTMLInputElement, IInputProps>(
  ({ hasError, label, helperText, ...props }, ref) => {
    const inputId = useId();

    return (
      <ContainerSC>
        <LabelSC htmlFor={inputId}>{label}</LabelSC>
        <InputSC hasError={hasError} id={inputId} ref={ref} {...props} />

        {hasError && <HelperTextSC>{helperText}</HelperTextSC>}
      </ContainerSC>
    );
  }
);

export default Input;
