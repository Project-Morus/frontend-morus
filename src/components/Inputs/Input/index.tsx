import { forwardRef } from "react";
import { ContainerSC, InputSC, LabelSC, ErrorTextSC } from "./styles";
import { IInputProps } from "./types";

const Input = forwardRef<HTMLInputElement, IInputProps>(
  ({ id, maxWidth, hasError, label, errorText, disabled, ...props }, ref) => {
    return (
      <ContainerSC $maxWidth={maxWidth}>
        <LabelSC $hasError={hasError} $disabled={disabled} htmlFor={id}>
          {label}
        </LabelSC>
        <InputSC disabled={disabled} $hasError={hasError} id={id} ref={ref} {...props} />

        {hasError && <ErrorTextSC>{errorText}</ErrorTextSC>}
      </ContainerSC>
    );
  }
);

export default Input;
