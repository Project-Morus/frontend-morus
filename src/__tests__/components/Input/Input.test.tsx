import { render, screen } from "@testing-library/react";
import Input from "../../../components/Input";
import "@testing-library/jest-dom";
import "jest-styled-components";
import { withTheme } from "../../../helpers/withTheme";

const InputWithTheme = withTheme(Input);

const idName = "idname";

describe("Input Component", () => {
  test("Verify if component is rendering", () => {
    render(<InputWithTheme label="Name" id={idName} />);

    const role = "textbox";

    expect(screen.getByRole(role)).toBeInTheDocument();
  });

  test("Verify if label is rendering correctly", () => {
    render(<InputWithTheme id={idName} label="Name" />);
    const labelElement = screen.getByText("Name");

    expect(labelElement).toBeInTheDocument();
  });

  test("Verify if error message is rendering correctly", () => {
    render(<InputWithTheme id={idName} hasError errorText="Error Message" label="Name" />);
    const errorMessage = screen.getByText("Error Message");

    expect(errorMessage).toBeInTheDocument();
  });

  test("Verify if input disabled is rendering correctly", () => {
    render(<InputWithTheme id={idName} disabled label="Name" />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toBeDisabled();
  });

  test("Snapshot Input", () => {
    const { asFragment } = render(<InputWithTheme id={idName} label="Name" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
