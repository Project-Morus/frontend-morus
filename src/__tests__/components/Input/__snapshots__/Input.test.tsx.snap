// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Snapshot Input 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.c0+.c0 {
  margin-top: 1rem;
}

.c2 {
  color: #4E4E4E;
  margin-bottom: 0.25rem;
}

.c3 {
  color: #4E4E4E;
  padding: 0.5rem;
  border: solid #BABABA 1px;
  border-radius: 5px;
}

.c3:focus {
  border-color: #0353A4;
  color: #0353A4;
}

.c3::placeholder {
  color: #BABABA;
}

.c3:placeholder .c3.c3 :disabled {
  color: #E9E9E9;
}

<div
    class="c0 c1"
  >
    <label
      class="c2"
      for="idname"
    >
      Name
    </label>
    <input
      class="c3"
      id="idname"
    />
  </div>
</DocumentFragment>
`;
